tsl/src/continuous_aggs/planner.c

```diff
@@ -355,3 +355,100 @@ constify_cagg_watermark(Query *parse)
+		return;
+
+	Cache *cache = ts_hypertable_cache_pin();
+
+	foreach (lc, parse->rtable)
+	{
+		RangeTblEntry *rte = lfirst(lc);
+
+		/*
+		 * Realtime cagg view will have 2 rtable entries, one for the materialized data and one for
+		 * the not yet materialized data.
+		 */
+		if (rte->rtekind != RTE_SUBQUERY || rte->relkind != RELKIND_VIEW ||
+			list_length(rte->subquery->rtable) != 2)
+			continue;
+
+		ContinuousAgg *cagg = ts_continuous_agg_find_by_relid(rte->relid);
+
+		/*
+		 * This optimization only applies to realtime caggs.
+		 */
+		if (!cagg || !cagg->data.finalized || cagg->data.materialized_only)
+			continue;
+
+		Hypertable *ht = ts_hypertable_cache_get_entry_by_id(cache, cagg->data.mat_hypertable_id);
+		Dimension const *dim = hyperspace_get_open_dimension(ht->space, 0);
+
+		/* We should only encounter hypertables with an open dimension */
+		if (!dim)
+			continue;
+
+		SortGroupClause *sort = linitial_node(SortGroupClause, parse->sortClause);
+		TargetEntry *tle = get_sortgroupref_tle(sort->tleSortGroupRef, parse->targetList);
+
+		/*
+		 * We only pushdown ORDER BY when it's single column
+		 * ORDER BY on the time column.
+		 */
+		AttrNumber time_col = dim->column_attno;
+		if (!IsA(tle->expr, Var) || castNode(Var, tle->expr)->varattno != time_col)
+			continue;
+
+		RangeTblEntry *mat_rte = linitial_node(RangeTblEntry, rte->subquery->rtable);
+		RangeTblEntry *rt_rte = lsecond_node(RangeTblEntry, rte->subquery->rtable);
+
+		mat_rte->subquery->sortClause = list_copy(parse->sortClause);
+		rt_rte->subquery->sortClause = list_copy(parse->sortClause);
+
+		TargetEntry *mat_tle = list_nth(mat_rte->subquery->targetList, time_col - 1);
+		TargetEntry *rt_tle = list_nth(rt_rte->subquery->targetList, time_col - 1);
+		linitial_node(SortGroupClause, mat_rte->subquery->sortClause)->tleSortGroupRef =
+			mat_tle->ressortgroupref;
+		linitial_node(SortGroupClause, rt_rte->subquery->sortClause)->tleSortGroupRef =
+			rt_tle->ressortgroupref;
+
+		SortGroupClause *cagg_group = linitial(rt_rte->subquery->groupClause);
+		cagg_group = list_nth(rt_rte->subquery->groupClause, rt_tle->ressortgroupref - 1);
+		cagg_group->sortop = sort->sortop;
+		cagg_group->nulls_first = sort->nulls_first;
+
+		Oid placeholder;
+		int16 strategy;
+		get_ordering_op_properties(sort->sortop, &placeholder, &placeholder, &strategy);
+
+		/*
+		 * If this is DESC order and the sortop is the commutator of the cagg_group sortop,
+		 * we can align the sortop of the cagg_group with the sortop of the sort clause, which
+		 * will allow us to have the GroupAggregate node to produce the correct order and avoid
+		 * having to resort.
+		 */
+		if (strategy == BTGreaterStrategyNumber)
+		{
+			/*
+			 * Try to align the sortop of the cagg_group with the sortop of the
+			 * sort clause.
+			 */
+			rte->subquery->rtable = list_make2(rt_rte, mat_rte);
+		}
+
+		parse->sortClause = NIL;
+		rte->subquery->sortClause = NIL;
```

Comment:
Do you prevent parallel append somehow? It's not going to work correctly with parallel, right?

Reply:
I get a feeling that it can't be done safely this way at a parse stage, we're making an assumption that the query will have a particular plan that we know has special properties that will make the result correct. But the modified query doesn't actually require these properties, so a different plan might be chosen, like the parallel one, and the result will be wrong.

Reply:
Maybe some path instead hook that turns MergeAppend into Append when the children don't intersect?

Reply:
Parallelism is now prevented when we do this optimization.